File: frontend/services/rpc.ts
import { API_BASE } from "./config";

export interface SendTransactionRequest {
  transaction: string; // base64 encoded signed transaction
}

export interface SendTransactionResponse {
  signature: string;
}

export interface ConfirmTransactionRequest {
  signature: string;
  blockhash: string;
  last_valid_block_height: number;
}

export interface ConfirmTransactionResponse {
  confirmed: boolean;
  error?: string;
}

/**
 * Send a signed transaction to Solana via the backend RPC proxy.
 * This avoids browser CORS/rate-limit issues with public RPC endpoints.
 */
export async function sendTransaction(
  transaction: string
): Promise<SendTransactionResponse> {
  const response = await fetch(`${API_BASE}/rpc/send-transaction`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ transaction }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.message || `HTTP ${response.status}`);
  }

  return response.json();
}

/**
 * Confirm a transaction on Solana via the backend RPC proxy.
 */
export async function confirmTransaction(
  request: ConfirmTransactionRequest
): Promise<ConfirmTransactionResponse> {
  const response = await fetch(`${API_BASE}/rpc/confirm-transaction`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.message || `HTTP ${response.status}`);
  }

  return response.json();
}

File: backend/src/api/handlers/rpc.rs
use std::sync::Arc;

use axum::{extract::State, Json};
use serde::{Deserialize, Serialize};

use crate::error::AppError;
use crate::AppState;

/// Send transaction request
#[derive(Debug, Deserialize)]
pub struct SendTransactionRequest {
    pub transaction: String, // base64 encoded signed transaction
}

/// Send transaction response
#[derive(Debug, Serialize)]
pub struct SendTransactionResponse {
    pub signature: String,
}

/// Confirm transaction request
#[derive(Debug, Deserialize)]
pub struct ConfirmTransactionRequest {
    pub signature: String,
    pub blockhash: String,
    pub last_valid_block_height: u64,
}

/// Confirm transaction response
#[derive(Debug, Serialize)]
pub struct ConfirmTransactionResponse {
    pub confirmed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Send a signed transaction to the Solana network
pub async fn send_transaction(
    State(state): State<Arc<AppState>>,
    Json(req): Json<SendTransactionRequest>,
) -> Result<Json<SendTransactionResponse>, AppError> {
    let signature = state.deposit.send_raw_transaction(&req.transaction).await?;
    Ok(Json(SendTransactionResponse { signature }))
}

/// Confirm a transaction on the Solana network
pub async fn confirm_transaction(
    State(state): State<Arc<AppState>>,
    Json(req): Json<ConfirmTransactionRequest>,
) -> Result<Json<ConfirmTransactionResponse>, AppError> {
    match state
        .deposit
        .confirm_transaction(&req.signature, &req.blockhash, req.last_valid_block_height)
        .await
    {
        Ok(confirmed) => Ok(Json(ConfirmTransactionResponse {
            confirmed,
            error: None,
        })),
        Err(e) => Ok(Json(ConfirmTransactionResponse {
            confirmed: false,
            error: Some(e.to_string()),
        })),
    }
}
